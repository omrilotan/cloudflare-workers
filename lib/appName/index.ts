export const appName = (url: URL): string => url.hostname.replace(/^www./, "");
