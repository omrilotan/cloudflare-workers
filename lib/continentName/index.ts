const codes = new Map<string, string>([
	["AF", "Africa"],
	["AN", "Antarctica"],
	["AS", "Asia"],
	["EU", "Europe"],
	["NA", "North America"],
	["OC", "Oceania"],
	["SA", "South America"],
]);

export const continentName = (code: string | undefined): string | undefined =>
	code && codes.get(code);
