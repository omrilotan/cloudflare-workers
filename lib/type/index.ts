/**
 * Get the type from Content-Type header
 */
export function type(header: string): string {
	try {
		const contentType = header.split(";")[0].trim().toLowerCase();
		const [type, subtype] = contentType.split("/");
		if (["image", "font", "video"].includes(type)) {
			return type;
		}
		if (subtype.endsWith("json")) {
			return "json";
		}
		if (subtype.includes("javascript")) {
			return "js";
		}
		if (subtype.includes("ecmascript")) {
			return "js";
		}
		if (subtype === "css") {
			return "css";
		}
		if (subtype === "plain") {
			return "text";
		}
		if (subtype.includes("html")) {
			return "html";
		}
		if (subtype.endsWith("xml")) {
			return "xml";
		}
		return "unknown";
	} catch (error) {
		return "unknown";
	}
}
