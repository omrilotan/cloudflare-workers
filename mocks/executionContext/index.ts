export class MockExecutionContext implements ExecutionContext {
	#promises: Promise<any>[] = [];
	props: Record<string, any> = {};
	waitUntil(promise) {
		this.#promises.push(promise);
	}
	passThroughOnException() {}
	async drain() {
		await Promise.all(this.#promises);
		this.#promises.length = 0;
	}
}
