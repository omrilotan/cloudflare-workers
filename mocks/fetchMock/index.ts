import { CacheStorageMock } from "../cachesMock";
const { fetch } = globalThis;

/**
 * A series of helper functions to mock and spy on the global fetch function
 */
export const fetchMock = {
	fetchCount: 0,

	promises: new Set<Promise<Response>>(),

	makePromise: async (promise: Promise<Response>): Promise<Response> => {
		fetchMock.promises.add(promise);
		const result = await promise;
		fetchMock.promises.delete(promise);
		return result;
	},

	/**
	 * Mount the global fetch with a Jest function
	 */
	mount: async (): Promise<void> => {
		const caches = new CacheStorageMock();
		const cache = await caches.open("fetchMockCache");
		(globalThis.fetch as jest.Mock) = jest.fn(
			async function (): Promise<Response> {
				let response;

				const [request, init] = arguments;
				const cf: RequestInitCfProperties = init?.cf || {};

				const cached = await fetchMock.makePromise(cache.match(request));
				if (cached) {
					response = cached;
				} else {
					response = await fetchMock.makePromise(fetch.apply(this, arguments));
					fetchMock.fetchCount++;
				}
				if (cf.cacheTtl > 0 && response.ok && response.status !== 206) {
					await cache.put(request, response);
				}
				return response;
			},
		);
	},

	/**
	 * Reset Counters
	 */
	reset: (): void => {
		fetchMock.fetchCount = 0;
	},

	/**
	 * Await any hanging promises
	 */
	drain: async (): Promise<void> => {
		await Promise.all(Array.from(fetchMock.promises));
	},

	/**
	 * Restore the global fetch
	 */
	unmount: async (): Promise<void> => {
		await fetchMock.drain();
		fetchMock.reset();
		globalThis.fetch = fetch;
	},
};
