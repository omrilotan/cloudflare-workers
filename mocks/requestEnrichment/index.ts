export const requestEnrichment = {
	mount: () => {
		Object.defineProperty<Request>(Request.prototype, "cf", {
			get: function (): IncomingRequestCfProperties {
				return {
					asn: 395747,
					asOrganization: "Google Cloud",
					botManagement: {
						corporateProxy: false,
						ja3Hash: "",
						score: 0,
						staticResource: false,
						verifiedBot: false,
					},
					clientAcceptEncoding: "gzip, deflate, br",
					clientTcpRtt: 0,
					colo: "DFW",
					country: "US",
					isEUCountry: null || "1",
					httpProtocol: "HTTP/2.0",
					requestPriority: "weight=192;exclusive=0;group=3;group-weight=127",
					tlsCipher: "AEAD-AES128-GCM-SHA256",
					tlsClientAuth: null,
					tlsVersion: "TLSv1.3",
					city: "Austin",
					continent: "NA",
					latitude: "30.27130",
					longitude: "-97.74260",
					postalCode: "78701",
					metroCode: "635",
					region: "Texas",
					regionCode: "TX",
					timezone: "America/Chicago",
				};
			},
			enumerable: true,
			configurable: true,
		});
	},
	unmount: () => {
		Object.defineProperty(Request.prototype, "cf", {
			value: null,
			writable: true,
			enumerable: true,
			configurable: true,
		});
	},
};
