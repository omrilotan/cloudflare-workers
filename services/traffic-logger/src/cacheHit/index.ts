export type CacheStatus =
	| "BYPASS"
	| "DEFERRED"
	| "DYNAMIC"
	| "EXPIRED"
	| "HIT"
	| "IGNORED"
	| "MISS"
	| "NONE"
	| "REVALIDATED"
	| "STALE"
	| "STREAM HIT"
	| "UPDATING";

export const cacheHit = (status: CacheStatus): 0 | 1 =>
	["HIT", "STREAM HIT", "REVALIDATED"].includes(status.toUpperCase()) ? 1 : 0;
