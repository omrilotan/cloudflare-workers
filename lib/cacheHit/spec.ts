import { cacheHit } from ".";
import type { CacheStatus } from ".";

describe("lib/cacheHit", () => {
	test.each(["HIT", "Revalidated", "stream hit"])(
		"%s is considered cache hit",
		(status) => {
			expect(cacheHit(status as CacheStatus)).toBe(1);
		}
	);
	test.each([
		"BYPASS",
		"DEFERRED",
		"DYNAMIC",
		"EXPIRED",
		"IGNORED",
		"MISS",
		"NONE",
		"STALE",
		"UPDATING",
		null,
		undefined,
	])("%s is considered cache miss", (status) => {
		expect(cacheHit(status as CacheStatus)).toBe(0);
	});
});
