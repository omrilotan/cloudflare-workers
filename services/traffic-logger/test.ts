import { CacheStatus } from "../../lib/cacheHit";
import { MockExecutionContext } from "../../mocks/executionContext";
import { fetchMock } from "../../mocks/fetchMock";
import { requestEnrichment } from "../../mocks/requestEnrichment";
import type { Env as AppEnv } from "./src/interfaces";
import type { Env as GatewayEnv } from "./src/gateway/index";

const log = jest.fn();
const discord = jest.fn();
let worker;
const ctx = new MockExecutionContext();
const UID_PATTERN =
	/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/;
const INDEX_PATTERN = /[0-9a-f]{32}/;

describe("gateway", () => {
	const env: GatewayEnv = {
		MAIN: {
			fetch: jest.fn(
				async (request: Request): Promise<Response> => new Response("ok")
			),
		},
		CANARY: {
			fetch: jest.fn(
				async (request: Request): Promise<Response> =>
					new Response("Server Error", { status: 500 })
			),
		},
		ROLLOUT: {
			get: jest.fn((key: string): string => "0"),
		} as unknown as KVNamespace,
		ROLLOUT_KEY: "rollout-key",
		ROLLOUT_HEADER: "rollout-header",
	};
	beforeAll(async () => {
		worker = (await import("./src/gateway")).default;
	});
	afterEach(() => jest.clearAllMocks());
	afterAll(() => jest.restoreAllMocks());
	test("default : main", async () => {
		const request = new Request("https://example.com");
		const response = await worker.fetch(request, env, ctx);
		expect(env.MAIN.fetch).toHaveBeenCalled();
		expect(env.CANARY.fetch).not.toHaveBeenCalled();
		expect(response.status).toBe(200);
	});
	test("rollout is on 100 : main", async () => {
		const request = new Request("https://example.com");
		(env.ROLLOUT.get as jest.Mock).mockImplementationOnce(() => "100");
		const response = await worker.fetch(request, env, ctx);
		expect(env.MAIN.fetch).not.toHaveBeenCalled();
		expect(env.CANARY.fetch).toHaveBeenCalled();
		expect(response.status).toBe(500);
	});
	test("rollout is on 50 : either", async () => {
		await Promise.all(
			Array.from({ length: 50 }).map(() => {
				const request = new Request("https://example.com");
				(env.ROLLOUT.get as jest.Mock).mockImplementationOnce(() => "50");
				return worker.fetch(request, env, ctx);
			})
		);
		const mainCalls = (env.MAIN.fetch as jest.Mock).mock.calls.length;
		const canaryCalls = (env.CANARY.fetch as jest.Mock).mock.calls.length;
		expect(mainCalls).toBeGreaterThan(0);
		expect(canaryCalls).toBeGreaterThan(0);
		expect(mainCalls + canaryCalls).toBe(50);
	});
	test("custom header : canary", async () => {
		const request = new Request("https://example.com", {
			headers: {
				[env.ROLLOUT_HEADER]: "true",
			},
		});
		const response = await worker.fetch(request, env, ctx);
		expect(env.MAIN.fetch).not.toHaveBeenCalled();
		expect(env.CANARY.fetch).toHaveBeenCalled();
		expect(response.status).toBe(500);
	});
});

describe("traffic-logger", (): void => {
	const env: AppEnv = {
		DISCORD_WEBHOOK: "https://discord.com/api/webhooks/1234567890/1234567890",
		LOGZIO_TOKEN: "xmC8duHhUaqqYoBWbMgGq1g6jxUJwtPG",
		VERSION: "a3b445d",
		RELEASE: "2022-10-11",
		SEND_ANALYTICS: true,
		SEND_LOGS: true,
		VARIATION: "test",
		TRAFFIC_ANALYTICS: {
			writeDataPoint: jest.fn(),
		},
	};
	beforeAll(async (): Promise<void> => {
		requestEnrichment.mount();
		await fetchMock.mount();
		(globalThis.fetch as jest.Mock).mockImplementation(
			async (): Promise<Response> => new Response("Ok", { status: 200 })
		);
		jest.mock("../../lib/log", () => ({ log }));
		jest.mock("../../lib/discord", () => ({ discord }));
		try {
			worker = (await import("./src/app")).default;
		} catch (e) {
			console.error(e);
		}
	});
	afterEach(async (): Promise<void> => {
		await fetchMock.drain();
		jest.clearAllMocks();
	});
	afterAll(async (): Promise<void> => {
		requestEnrichment.unmount();
		await fetchMock.unmount();
		jest.restoreAllMocks();
	});

	test.each([
		[
			"logs traffic for",
			"https://httpbin.org/index.html",
			"MISS",
			"text/html",
			true,
		],
		[
			"does not log traffic for",
			"https://httpbin.org/script.239874.js",
			"HIT",
			"application/javascript",
			false,
		],
	])(
		"%s %s",
		async (
			description: string,
			url: string,
			cacheStatus: CacheStatus,
			contentType: string,
			shouldLog: boolean
		): Promise<void> => {
			(globalThis.fetch as jest.Mock).mockImplementationOnce(
				async () =>
					new Response("Ok", {
						status: 200,
						headers: new Headers([
							["CF-Cache-Status", cacheStatus],
							["Content-Type", contentType],
							["ETag", "u6LbP7BGEntOvnXR3SchEDg7KX2WW72Q"],
							["X-Powered-By", "Fiber"],
						]),
					})
			);
			const request = new Request(url, {
				method: "GET",
				headers: new Headers([
					[
						"Accept",
						"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
					],
					["CF-Connecting-IP", "203.0.113.1"],
					["Referer", "https://www.duckduckgo.com/"],
					[
						"Sec-CH-UA",
						'"Google Chrome";v="105", "Not)A;Brand";v="8", "Chromium";v="105"',
					],
					[
						"User-Agent",
						"Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:106.0) Gecko/20100101 Firefox/106.0",
					],
				]),
			});
			const response = await worker.fetch(request, env, ctx);

			expect(log).toHaveBeenCalledTimes(shouldLog ? 1 : 0);
			if (shouldLog) {
				const [[type, record, token]] = log.mock.calls;
				expect(type).toBe("traffic");
				expect(token).toBe(env.LOGZIO_TOKEN);
				expect(record.request_id).toMatch(UID_PATTERN);
				expect(record.duration).toBeGreaterThan(-1);
				delete record.request_id;
				delete record.duration;
				expect(record).toMatchSnapshot();
				const [[analytics_record]] = (
					env.TRAFFIC_ANALYTICS.writeDataPoint as jest.Mock
				).mock.calls;
				expect(analytics_record.indexes.pop()).toMatch(INDEX_PATTERN);
				const duration = analytics_record.doubles.splice(1, 1)[0];
				expect(duration).toBeGreaterThan(-1);
				expect(analytics_record).toMatchSnapshot();
			}

			expect(response.status).toBe(200);
			expect(response.headers.get("server-timing")).toMatch(
				new RegExp(
					`CDN-Origin-Fetch; dur=\\d+; desc="origin rtt",\\s*Cache-Status; dur=[01]; desc="${cacheStatus}"`
				)
			);
			response.headers.delete("server-timing");
			expect(Array.from(response.headers)).toMatchSnapshot();
		}
	);
});
