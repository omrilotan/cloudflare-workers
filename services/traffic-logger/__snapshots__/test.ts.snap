// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traffic-logger logs traffic for https://httpbin.org/index.html 1`] = `
{
  "app": "httpbin.org",
  "as": "Google Cloud, 395747",
  "automated": true,
  "browser": "Google Chrome 105",
  "cache_status": "MISS",
  "content_type": "html",
  "ip": "203.0.113.1",
  "level": "info",
  "location": "Austin, United States, North America",
  "logger": "a3b445d",
  "message": "200 "https://httpbin.org/index.html" from Austin, United States, North America",
  "referrer": "https://www.duckduckgo.com/",
  "request": "GET https://httpbin.org/index.html",
  "status": 200,
}
`;

exports[`traffic-logger logs traffic for https://httpbin.org/index.html 2`] = `
[
  [
    "accept-ch",
    "UA,Sec-CH-UA,Sec-CH-UA-Mobile,Sec-CH-UA-Full-Version,Sec-CH-UA-Full-Version-List,Sec-CH-UA-Model,Sec-CH-UA-Platform,Sec-CH-UA-Platform-Version,ECT",
  ],
  [
    "app-version",
    "a3b445d",
  ],
  [
    "cf-cache-status",
    "MISS",
  ],
  [
    "content-type",
    "text/html",
  ],
  [
    "strict-transport-security",
    "max-age=31536000; includeSubDomains",
  ],
  [
    "timing-allow-origin",
    "*",
  ],
  [
    "vary",
    "Accept, Accept-Encoding, User-Agent",
  ],
]
`;

exports[`traffic-logger logs traffic for https://httpbin.org/script.239874.js 1`] = `
[
  [
    "accept-ch",
    "UA,Sec-CH-UA,Sec-CH-UA-Mobile,Sec-CH-UA-Full-Version,Sec-CH-UA-Full-Version-List,Sec-CH-UA-Model,Sec-CH-UA-Platform,Sec-CH-UA-Platform-Version,ECT",
  ],
  [
    "app-version",
    "a3b445d",
  ],
  [
    "cf-cache-status",
    "HIT",
  ],
  [
    "content-type",
    "application/javascript",
  ],
  [
    "strict-transport-security",
    "max-age=31536000; includeSubDomains",
  ],
  [
    "timing-allow-origin",
    "*",
  ],
  [
    "vary",
    "Accept, Accept-Encoding, User-Agent",
  ],
]
`;
